### PR TITLE
Document the use of mandatory macro parameters and add named macro tokens

### PR DESCRIPTION
**Pull request by bernt-matthias**
learned while working on https://github.com/galaxyproject/galaxy/pull/13152

### CODE REVIEW

File: docs/_writing_macros.rst [hunk 1]
```diff
@@ -253,27 +253,35 @@ Parameterizing XML Macros (with tokens)
 -----------------------------------------
 
 In addition to using ``yield`` blocks, there is another way to parameterize
-macros through the use of specifying ``token_xyz`` attributes on the macro
+macros through the use of specifying 
+
+- required parameters as comma separated list of parameter names using the
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:258]:
```suggestion
- required parameters as comma-separated list of parameter names using the
```

File: docs/_writing_macros.rst [hunk 2]
```diff
@@ -253,27 +253,35 @@ Parameterizing XML Macros (with tokens)
...
+``tokens`` attribute of the macro, e.g. ``tokens="foo,bar"`` and then using
+@FOO@ and @BAR@ anywhere in the macro 
+- optional parameters as ``token_xyz`` attributes on the macro
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:261]:
```suggestion
- optional parameters as ``token_xyz="default_value"`` attributes of the ``xml`` 
```

File: docs/_writing_macros.rst [hunk 3]
```diff
@@ -253,27 +253,35 @@ Parameterizing XML Macros (with tokens)
...
 
 In addition to using ``yield`` blocks, there is another way to parameterize
-macros through the use of specifying ``token_xyz`` attributes on the macro
+macros through the use of specifying 
+
+- required parameters as comma separated list of parameter names using the
+``tokens`` attribute of the macro, e.g. ``tokens="foo,bar"`` and then using
+@FOO@ and @BAR@ anywhere in the macro 
+- optional parameters as ``token_xyz`` attributes on the macro 
 definition, and then using ``@XYZ@`` anywhere within the XML.
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:262]:
```suggestion
  element, and then using ``@XYZ@`` anywhere inside the macro definition.
```

**Review by nsoranzo — APPROVED** [review]

File: docs/_writing_macros.rst [hunk 4]
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
...
+          <option value="b">B</option>
+          <yield name="more_options"/>
+        </param>
+        <when value="a">
+          <param type="select">
+            <yield />
+          </param>
+        </when>
+        <when value="b">
+          <param ... />
```
Comment (mvdbeek) [inline docs/_writing_macros.rst:270]: I think it helps to have this be somewhat valid XML.

File: docs/_writing_macros.rst [hunk 5]
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
           </options>
         </expand>
 
+From Galaxy release 22.01 named yields are supported. That is, if the macro contains
+``<yield name="xyz"/>`` it is replaced by the content of the ``token` tag with the same name.
+Token tags need to be direct children of the ``expand`` element.
+
```
Comment (mvdbeek) [inline docs/_writing_macros.rst:254]: Can we have the example be something like:
Given the following macro:
...
and expanding the macro in the following way:
...
we get the following expanded definition:
...
?

File: docs/_writing_macros.rst [hunk 6]
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
...
+        <when value="b">
+          <param ... />
+        </when>
+        <yield name="more_whens">
+      </conditional>
+    </xml>
+
+.. code-block:: xml
+
+        <expand macro="named_yields_example">
```
Comment (mvdbeek) [inline docs/_writing_macros.rst:278]: The indentiation looks weird to me, why not start on the left side ?

File: docs/_writing_macros.rst [hunk 7]
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
...
+            <when value="c">
+              <param type="select">
+                <yield />
+              </param>
+            </when>
+          </token>
+          <options from_data_table="tophat2_indexes">
+            <filter type="sort_by" column="2"/>
+            <validator type="no_options" message="No genomes are available for the selected input dataset"/>
+          </options>
```
Comment (mvdbeek) [inline docs/_writing_macros.rst:292]:
```suggestion
```
This is unrelated to the macro example, right ?

File: docs/_writing_macros.rst [hunk 8]
```diff
@@ -248,32 +248,94 @@ The following expand declarations have replaced the original conditional element
           </options>
         </expand>
 
+From Galaxy release 22.01 named yields are supported. That is, if the macro contains
+``<yield name="xyz"/>`` it is replaced by the content of the ``token` tag with the same name.
+Token tags need to be direct children of the ``expand`` element.
```
Comment (mvdbeek) [inline docs/_writing_macros.rst:253]:
```suggestion
Token tags need to be direct children of the ``expand`` element.
This is useful when you would like to compose options of a select list, as shown in the example below.
```

File: docs/_writing_macros.rst (hunk 7 above)
Comment (bernt-matthias) [reply docs/_writing_macros.rst:292]: The unnamed yield is replaced by this.

Comment (mvdbeek) [reply docs/_writing_macros.rst:292]: Oh, I have not understood this. Why is that ? How do I read this as a tool author ?

Comment (bernt-matthias) [reply docs/_writing_macros.rst:292]: Essentially as before: The state was that the content of the expand tag is inserted at the place(s) of the yield tags in the macro. 

In order to keep backward compatibility I implemented it such still all content of the expand minus the token tags and their content is used for the replacement.

Comment (mvdbeek) [reply docs/_writing_macros.rst:292]: Ah, I see, yeah, that makes sense.

**Comment by bernt-matthias** [comment]
I hope that I have addressed all comment. 

Just had the idea that one could allow content to `yield` tags and interpret it as default if the corresponding token is not given. 
Seems that writing docs is not only annoying but may result in better ideas and helps to reason about code.

**Comment by bernt-matthias** [comment]
Hey @nsoranzo there were quite a few commits after your approval. Are you still OK with merging?

File: docs/_writing_macros.rst [hunk 9]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
+
+   <expand macro="refGenomeSourceConditional">
+     <options from_data_table="tophat2_indexes">
+       <filter type="sort_by" column="2"/>
+       <validator type="no_options" message="No genomes are available for the selected input dataset"/>
+     </options>
+   </expand>
+
+From Galaxy release 22.01 named yields are supported. That is, if the macro contains
+``<yield name="xyz"/>`` it is replaced by the content of the ``token`` tag with the same name.
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:252]:
```suggestion
``<yield name="xyz"/>`` it is replaced by the content of the ``token`` element with the same name.
```

File: docs/_writing_macros.rst [hunk 10]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
+Token tags need to be direct children of the ``expand`` element. This is useful if different
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:253]:
```suggestion
Token elements need to be direct children of the ``expand`` element. This is useful if different
```

File: docs/_writing_macros.rst [hunk 11]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
+     <options from_data_table="tophat2_indexes">
+       <filter type="sort_by" column="2"/>
+       <validator type="no_options" message="No genomes are available for the selected input dataset"/>
+     </options>
+   </expand>
+
+we get the following expanded definition:
+
+.. code-block:: xml
+   :emphasize-lines: 6,9-14,17-24
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:305]:
```suggestion
   :emphasize-lines: 6,10-13,17-24
```

File: docs/_writing_macros.rst [hunk 12]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
+             <filter type="sort_by" column="2"/>
+             <validator type="no_options" message="No genomes are available for the selected input dataset"/>
+           </options>
+         </param>
+       </when>
+     </conditional>
+   </xml>
+
+Named yields are replaced in the order of the tokens defined in the ``expand``
+tag. Unamed yields are replaced after all named tokens have been replaced (by
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:335]:
```suggestion
tag. Unnamed yields are replaced after all named tokens have been replaced (by
```

File: docs/_writing_macros.rst [hunk 13]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
+the non-token child elements of the expand tag). If there are named
+yields that have no corresponding token, then they are treated like unnamed
+yields. Note that, unnamed and named tokens can be used multiple times in a
+macro, then each occurence is preplaced by the corresponding content definedin
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:339]:
```suggestion
macro, then each occurrence is replaced by the corresponding content defined in
```

File: docs/_writing_macros.rst [hunk 14]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
+         </param>
+       </when>
+     </conditional>
+   </xml>
+
+Named yields are replaced in the order of the tokens defined in the ``expand``
+tag. Unamed yields are replaced after all named tokens have been replaced (by
+the non-token child elements of the expand tag). If there are named
+yields that have no corresponding token, then they are treated like unnamed
+yields. Note that, unnamed and named tokens can be used multiple times in a
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:338]:
```suggestion
yields. Note that unnamed and named tokens can be used multiple times in a
```

File: docs/_writing_macros.rst [hunk 15]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
+Further, note that the order of the replacements offers some possibilities
+to achieve recursion-like replacements, since a token may contain further named
+or unnamed ``yield`` tags (see for instance the ``yield`` tag contained in the
+named token ``more_whens``).
 
 -----------------------------------------
 Parameterizing XML Macros (with tokens)
 -----------------------------------------
 
 In addition to using ``yield`` blocks, there is another way to parameterize
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:351]:
```suggestion
In addition to using ``yield`` blocks, there is another way to parametrize
```

File: docs/_writing_macros.rst [hunk 16]
```diff
@@ -163,124 +163,244 @@ For instance, previously the tophat wrapper contained the following definition:
...
 
 .. code-block:: xml
 
-    <inputs>
-        <param name="myvar" type="color" label="Pick a color" value="#ff0000" />
-        <param name="c2" type="color" label="Choose a different color" value="#0000ff" />
-    </inputs>
+   <macros>
+     <xml name="color" tokens="attr,attr_value" token_quote="__" token_label="label">
+         <param __VAR__="__ATTR_VALUE__" label="__LABEL__"/>
```
Comment (nsoranzo) [inline docs/_writing_macros.rst:399]:
```suggestion
         <param __ATTR__="__ATTR_VALUE__" label="__LABEL__"/>
```

**Review by nsoranzo — APPROVED** [review]
